### PR TITLE
usbdev/mtp: support media transfer protocol (MTP) class driver

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
usbdev/mtp: support mtp class driver
## Impact
support media transfer protocol (MTP) class driver
## Testing
daily test

### CODE REVIEW

File: drivers/usbdev/Kconfig [hunk 1]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+	int "Interrupt IN endpoint number"
+	default 5
+	---help---
+		The logical 7-bit address of a hardware endpoint that supports
+		interrupt IN operation.  Default: 5
+
+endif
+
+config USBMTP_EPBULKOUT_FSSIZE
+	int "Bulk OUT full speed MAXPACKET size"
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1292]:
```suggestion
	int "Bulk OUT Full Speed MAXPACKET size"

File: drivers/usbdev/Kconfig [hunk 2]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+	default 64
+	---help---
+		Max package size for the bulk OUT endpoint if full speed mode.
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1295]:
```suggestion
		Max package size for the bulk OUT endpoint if Full Speed mode.

File: drivers/usbdev/Kconfig [hunk 3]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+		Default 64.
+
+if USBDEV_DUALSPEED
+
+config USBMTP_EPBULKOUT_HSSIZE
+	int "Bulk OUT out high speed MAXPACKET size"
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1301]:
```suggestion
	int "Bulk OUT out High Speed MAXPACKET size"

File: drivers/usbdev/Kconfig [hunk 4]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+	default 512
+	---help---
+		Max package size for the bulk OUT endpoint if high speed mode.
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1304]:
```suggestion
		Max package size for the bulk OUT endpoint if High Speed mode.

File: drivers/usbdev/Kconfig [hunk 5]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+		Default 512.
+
+endif # USBDEV_DUALSPEED
+
+config USBMTP_EPBULKIN_FSSIZE
+	int "Bulk IN full speed MAXPACKET size"
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1310]: int "Bulk IN Full Speed MAXPACKET size"

File: drivers/usbdev/Kconfig [hunk 6]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+	default 64
+	---help---
+		Max package size for the bulk IN endpoint if full speed mode.
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1313]:
```suggestion
		Max package size for the bulk IN endpoint if Full Speed mode.

File: drivers/usbdev/Kconfig [hunk 7]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+		Default 64.
+
+if USBDEV_DUALSPEED
+
+config USBMTP_EPBULKIN_HSSIZE
+	int "Bulk IN high speed MAXPACKET size"
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1319]:
```suggestion
	int "Bulk IN High Speed MAXPACKET size"

File: drivers/usbdev/Kconfig [hunk 8]
```diff
@@ -1234,4 +1234,157 @@ config USBDEV_FS_NPOLLWAITERS
...
+	default 512
+	---help---
+		Max package size for the bulk IN endpoint if high speed mode.
```
Comment (acassis) [inline drivers/usbdev/Kconfig:1322]:
```suggestion
		Max package size for the bulk IN endpoint if High Speed mode.

File: drivers/usbdev/mtp.c [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+#else
+#  define USBMTP_REMOTEWAKEUP      (0)
+#endif
+
+/* Buffer big enough for any of our descriptors (the config descriptor is the
+ * biggest).
+ */
+
+#define USBMTP_MXDESCLEN           (64)
+#define USBMTP_MAXSTRLEN           (USBMTP_MXDESCLEN-2)
```
Comment (acassis) [inline drivers/usbdev/mtp.c:75]:
```suggestion
#define USBMTP_MAXSTRLEN           (USBMTP_MXDESCLEN - 2)

File: drivers/usbdev/mtp.c [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+
+/* Device descriptor values */
+
+#define USBMTP_VERSIONNO           (0x0101) /* Device version number 1.1 (BCD) */
+
+/* String language */
+
+#define USBMTP_STR_LANGUAGE        (0x0409) /* en-us */
+
+/* Descriptor strings.  If there serial device is part of a composite device
```
Comment (acassis) [inline drivers/usbdev/mtp.c:85]:
```suggestion
/* Descriptor strings.  If this MTP device is part of a composite device

Comment (Donny9) [reply drivers/usbdev/mtp.c]: Done~

File: drivers/usbdev/mtp.c (hunk 9 above)
Comment (Donny9) [reply drivers/usbdev/mtp.c]: Done~

File: drivers/usbdev/Kconfig (hunk 8 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~

File: drivers/usbdev/Kconfig (hunk 7 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~

File: drivers/usbdev/Kconfig (hunk 6 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~

File: drivers/usbdev/Kconfig (hunk 5 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~

File: drivers/usbdev/Kconfig (hunk 4 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~

File: drivers/usbdev/Kconfig (hunk 3 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~

File: drivers/usbdev/Kconfig (hunk 2 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~

File: drivers/usbdev/Kconfig (hunk 1 above)
Comment (Donny9) [reply drivers/usbdev/Kconfig]: Done~ Thanks

**Comment by davids5** [comment]
@Donny9 What HW was this tested on? Please provide a config for testing.

**Comment by Donny9** [comment]
> @Donny9 What HW was this tested on? Please provide a config for testing.

@davids5  I tested on the simulator base sim:usbdev, but it relies on an mtp responder service. If only have this patch, you can enumerate normally through dummy hcd.
enable CONFIG_USBMTP=y and disable CONFIG_USBADB
ubuntu dmesg log:
```
djz:nuttx$ sudo dmesg 
[1169256.505256] usb 3-1: USB disconnect, device number 22
[1169258.731447] usb 3-1: new high-speed USB device number 23 using dummy_hcd
[1169259.047409] usb 3-1: New USB device found, idVendor=18d1, idProduct=4e11, bcdDevice= 1.01
[1169259.047424] usb 3-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[1169259.047431] usb 3-1: Product: Nuttx MTP
[1169259.047436] usb 3-1: Manufacturer: NuttX
[1169259.047440] usb 3-1: SerialNumber: 1234
```

image

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**